### ISSUE TITLE
scala.reflect.api.Trees MatchError on using combinator parsers with Scala 2.13.6

### ISSUE THREAD

**Issue by Pyppe** [scala/bug#12458]
## reproduction steps

using Scala `2.13.6`,

```scala
val NP_ATTR_CONJ = (DET*) ~ (((ATTR+) <~ COMMA)*) ~ (ATTR+) ~ CONJ ~ (ATTR+) ~ (NP_HEAD+) ^^ {
  case det ~ attrs ~ attr1 ~ conj ~ attr2 ~ nphead => NPPhrase_DET_ATTR_CONJ_N(det, attrs.flatten, attr1, conj, attr2, nphead.dropRight(1), nphead.last)
}
```

## problem

Crashes the compiler:

```
scala.MatchError: Parens(List(ATTR.$plus.$less$tilde(COMMA))) (of class scala.tools.nsc.ast.Trees$Parens)
scala.reflect.api.Trees.xtransform(Trees.scala:2654)
```

## details

The code worked fine in Scala `2.13.4`. We fixed the issue for 2.13.6 by adding dots:
```scala
val NP_ATTR_CONJ = (DET.*) ~ (((ATTR.+) <~ COMMA).*) ~ (ATTR.+) ~ CONJ ~ (ATTR.+) ~ (NP_HEAD.+) ^^ {
  case det ~ attrs ~ attr1 ~ conj ~ attr2 ~ nphead => NPPhrase_DET_ATTR_CONJ_N(det, attrs.flatten, attr1, conj, attr2, nphead.dropRight(1), nphead.last)
}
```

**Comment by lrytz** [comment]
Could you provide a full running example?

I ran `scalac Test.scala -Vprint:parser -Yshow-trees -Ystop-after:parser` with 2.13.6 and 2.13.4 on your code and the output is identical, I can't see any `Parens` trees.

**Comment by som-snytt** [comment]
https://github.com/scala/bug/issues/12415 perhaps

**Comment by Pyppe** [comment]
I could not reproduce with `scalac` either. I am assuming it's related to our sbt scalac options then:
```
"-encoding", "UTF-8",
"-deprecation",
"-unchecked",
"-feature",
"-Xlint:-adapted-args,-package-object-classes",
"-Xsource:3",
"-Ypatmat-exhaust-depth", "320",
"-Ymacro-annotations",
"-Werror",
"-Wdead-code",
"-Wunused",
"-Wvalue-discard",
"-Wextra-implicit",
"-Wnumeric-widen"
```

**Comment by SethTisue** [comment]
The next steps here are:

* Provide a self-contained reproduction case. We cannot compile your code, it refers to definitions you haven't provided.
* Experiment to determine what the responsible compiler flag is (as Som indicates, `-Xsource:3` is a likely culprit).

**Comment by Pyppe** [comment]
### Source code: Example.scala
```scala
package pyppe.example

import scala.util.parsing.combinator.{PackratParsers, Parsers}

object Example extends Parsers with PackratParsers {
  val NP_ATTR_CONJ = (DET*) ~ (((ATTR+) <~ COMMA)*) ~ (ATTR+) ~ CONJ ~ (ATTR+) ~ (NP_HEAD+) ^^ {
    case det ~ attrs ~ attr1 ~ conj ~ attr2 ~ nphead => NPPhrase_DET_ATTR_CONJ_N(det, attrs.flatten, attr1, conj, attr2, nphead.dropRight(1), nphead.last)
  }
}
```

### Compilation

The problem seems to arise while using both `-Xsource:3` and  `-Ymacro-annotations` in combination:

```shell
scalac -Xsource:3 -Ymacro-annotations Example.scala
```

The `Example.scala` references variables not available in the example, but because the compiler throws an error I assume this can be tested even without the definitions?

### Output
```
error: scala.MatchError: Parens(List(ATTR.$plus.$less$tilde(COMMA))) (of class scala.tools.nsc.ast.Trees$Parens)
        at scala.reflect.api.Trees.xtransform(Trees.scala:2654)
        at scala.reflect.api.Trees.xtransform$(Trees.scala:2654)
        at scala.tools.nsc.Global.scala$tools$nsc$ast$Trees$$super$xtransform(Global.scala:45)
        at scala.tools.nsc.ast.Trees.xtransform(Trees.scala:208)
        at scala.tools.nsc.ast.Trees.xtransform$(Trees.scala:197)
        at scala.tools.nsc.Global.xtransform(Global.scala:45)
        at scala.tools.nsc.Global.xtransform(Global.scala:45)
        at scala.reflect.internal.Trees$TreeContextApiImpl.transform(Trees.scala:277)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.internal.Trees$Typed.transform(Trees.scala:743)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.api.Trees$Transformer.transformTrees(Trees.scala:2591)
        at scala.reflect.internal.Trees$Apply.transform(Trees.scala:786)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.internal.Trees$Select.transform(Trees.scala:852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.internal.Trees$Apply.transform(Trees.scala:786)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.internal.Trees$Select.transform(Trees.scala:852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.internal.Trees$Apply.transform(Trees.scala:786)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.internal.Trees$Select.transform(Trees.scala:852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.internal.Trees$Apply.transform(Trees.scala:786)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.internal.Trees$Select.transform(Trees.scala:852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.internal.Trees$Apply.transform(Trees.scala:786)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.internal.Trees$Select.transform(Trees.scala:852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.internal.Trees$Apply.transform(Trees.scala:786)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.internal.Trees$ValDef.$anonfun$transform$4(Trees.scala:433)
        at scala.reflect.api.Trees$Transformer.atOwner(Trees.scala:2633)
        at scala.reflect.internal.Trees$ValDef.transform(Trees.scala:432)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.api.Trees$Transformer.$anonfun$transformStats$1(Trees.scala:2622)
        at scala.reflect.api.Trees$Transformer.transformStats(Trees.scala:2620)
        at scala.reflect.internal.Trees$Template.transform(Trees.scala:544)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.api.Trees$Transformer.transformTemplate(Trees.scala:2595)
        at scala.reflect.internal.Trees$ModuleDef.$anonfun$transform$3(Trees.scala:396)
        at scala.reflect.api.Trees$Transformer.atOwner(Trees.scala:2633)
        at scala.reflect.internal.Trees$ModuleDef.transform(Trees.scala:395)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.api.Trees$Transformer.$anonfun$transformStats$1(Trees.scala:2622)
        at scala.reflect.api.Trees$Transformer.transformStats(Trees.scala:2620)
        at scala.reflect.internal.Trees$PackageDef.$anonfun$transform$1(Trees.scala:342)
        at scala.reflect.api.Trees$Transformer.atOwner(Trees.scala:2633)
        at scala.reflect.internal.Trees$PackageDef.transform(Trees.scala:342)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees.duplicateAndKeepPositions(Trees.scala:1884)
        at scala.reflect.internal.Trees.duplicateAndKeepPositions$(Trees.scala:1884)
        at scala.reflect.internal.SymbolTable.duplicateAndKeepPositions(SymbolTable.scala:28)
        at scala.tools.nsc.typechecker.MacroAnnotationAttachments.attachSource(StdAttachments.scala:235)
        at scala.tools.nsc.typechecker.MacroAnnotationAttachments.attachSource$(StdAttachments.scala:235)
        at scala.tools.nsc.Global$$anon$5.attachSource(Global.scala:484)
        at scala.tools.nsc.typechecker.MacroAnnotationNamers$MacroAnnotationNamer.deriveSymbolFromSource$1(MacroAnnotationNamers.scala:76)
        at scala.tools.nsc.typechecker.MacroAnnotationNamers$MacroAnnotationNamer.createAssignAndEnterSymbol(MacroAnnotationNamers.scala:79)
        at scala.tools.nsc.typechecker.MacroAnnotationNamers$MacroAnnotationNamer.dispatch$1(MacroAnnotationNamers.scala:40)
        at scala.tools.nsc.typechecker.MacroAnnotationNamers$MacroAnnotationNamer.standardEnterSym(MacroAnnotationNamers.scala:47)
        at scala.tools.nsc.typechecker.AnalyzerPlugins.pluginsEnterSym(AnalyzerPlugins.scala:495)
        at scala.tools.nsc.typechecker.AnalyzerPlugins.pluginsEnterSym$(AnalyzerPlugins.scala:494)
        at scala.tools.nsc.Global$$anon$5.pluginsEnterSym(Global.scala:484)
        at scala.tools.nsc.typechecker.Namers$Namer.enterSym(Namers.scala:287)
        at scala.tools.nsc.typechecker.Analyzer$namerFactory$$anon$1.apply(Analyzer.scala:48)
        at scala.tools.nsc.Global$GlobalPhase.applyPhase(Global.scala:454)
        at scala.tools.nsc.Global$GlobalPhase.run(Global.scala:401)
        at scala.tools.nsc.Global$Run.compileUnitsInternal(Global.scala:1519)
        at scala.tools.nsc.Global$Run.compileUnits(Global.scala:1503)
        at scala.tools.nsc.Global$Run.compileSources(Global.scala:1495)
        at scala.tools.nsc.Global$Run.compile(Global.scala:1630)
        at scala.tools.nsc.Driver.doCompile(Driver.scala:48)
        at scala.tools.nsc.MainClass.doCompile(Main.scala:30)
        at scala.tools.nsc.Driver.process(Driver.scala:68)
        at scala.tools.nsc.Driver.main(Driver.scala:82)
        at scala.tools.nsc.Main.main(Main.scala)
Exception in thread "main" scala.MatchError: Parens(List(ATTR.$plus.$less$tilde(COMMA))) (of class scala.tools.nsc.ast.Trees$Parens)
        at scala.reflect.api.Trees.xtransform(Trees.scala:2654)
        at scala.reflect.api.Trees.xtransform$(Trees.scala:2654)
        at scala.tools.nsc.Global.scala$tools$nsc$ast$Trees$$super$xtransform(Global.scala:45)
        at scala.tools.nsc.ast.Trees.xtransform(Trees.scala:208)
        at scala.tools.nsc.ast.Trees.xtransform$(Trees.scala:197)
        at scala.tools.nsc.Global.xtransform(Global.scala:45)
        at scala.tools.nsc.Global.xtransform(Global.scala:45)
        at scala.reflect.internal.Trees$TreeContextApiImpl.transform(Trees.scala:277)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.internal.Trees$Typed.transform(Trees.scala:743)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.api.Trees$Transformer.transformTrees(Trees.scala:2591)
        at scala.reflect.internal.Trees$Apply.transform(Trees.scala:786)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.internal.Trees$Select.transform(Trees.scala:852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.internal.Trees$Apply.transform(Trees.scala:786)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.internal.Trees$Select.transform(Trees.scala:852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.internal.Trees$Apply.transform(Trees.scala:786)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.internal.Trees$Select.transform(Trees.scala:852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.internal.Trees$Apply.transform(Trees.scala:786)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.internal.Trees$Select.transform(Trees.scala:852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.internal.Trees$Apply.transform(Trees.scala:786)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.internal.Trees$Select.transform(Trees.scala:852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.internal.Trees$Apply.transform(Trees.scala:786)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.internal.Trees$ValDef.$anonfun$transform$4(Trees.scala:433)
        at scala.reflect.api.Trees$Transformer.atOwner(Trees.scala:2633)
        at scala.reflect.internal.Trees$ValDef.transform(Trees.scala:432)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.api.Trees$Transformer.$anonfun$transformStats$1(Trees.scala:2622)
        at scala.reflect.api.Trees$Transformer.transformStats(Trees.scala:2620)
        at scala.reflect.internal.Trees$Template.transform(Trees.scala:544)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.api.Trees$Transformer.transformTemplate(Trees.scala:2595)
        at scala.reflect.internal.Trees$ModuleDef.$anonfun$transform$3(Trees.scala:396)
        at scala.reflect.api.Trees$Transformer.atOwner(Trees.scala:2633)
        at scala.reflect.internal.Trees$ModuleDef.transform(Trees.scala:395)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1849)
        at scala.reflect.api.Trees$Transformer.$anonfun$transformStats$1(Trees.scala:2622)
        at scala.reflect.api.Trees$Transformer.transformStats(Trees.scala:2620)
        at scala.reflect.internal.Trees$PackageDef.$anonfun$transform$1(Trees.scala:342)
        at scala.reflect.api.Trees$Transformer.atOwner(Trees.scala:2633)
        at scala.reflect.internal.Trees$PackageDef.transform(Trees.scala:342)
        at scala.reflect.internal.Trees$Duplicator.transform(Trees.scala:1852)
        at scala.reflect.internal.Trees.duplicateAndKeepPositions(Trees.scala:1884)
        at scala.reflect.internal.Trees.duplicateAndKeepPositions$(Trees.scala:1884)
        at scala.reflect.internal.SymbolTable.duplicateAndKeepPositions(SymbolTable.scala:28)
        at scala.tools.nsc.typechecker.MacroAnnotationAttachments.attachSource(StdAttachments.scala:235)
        at scala.tools.nsc.typechecker.MacroAnnotationAttachments.attachSource$(StdAttachments.scala:235)
        at scala.tools.nsc.Global$$anon$5.attachSource(Global.scala:484)
        at scala.tools.nsc.typechecker.MacroAnnotationNamers$MacroAnnotationNamer.deriveSymbolFromSource$1(MacroAnnotationNamers.scala:76)
        at scala.tools.nsc.typechecker.MacroAnnotationNamers$MacroAnnotationNamer.createAssignAndEnterSymbol(MacroAnnotationNamers.scala:79)
        at scala.tools.nsc.typechecker.MacroAnnotationNamers$MacroAnnotationNamer.dispatch$1(MacroAnnotationNamers.scala:40)
        at scala.tools.nsc.typechecker.MacroAnnotationNamers$MacroAnnotationNamer.standardEnterSym(MacroAnnotationNamers.scala:47)
        at scala.tools.nsc.typechecker.AnalyzerPlugins.pluginsEnterSym(AnalyzerPlugins.scala:495)
        at scala.tools.nsc.typechecker.AnalyzerPlugins.pluginsEnterSym$(AnalyzerPlugins.scala:494)
        at scala.tools.nsc.Global$$anon$5.pluginsEnterSym(Global.scala:484)
        at scala.tools.nsc.typechecker.Namers$Namer.enterSym(Namers.scala:287)
        at scala.tools.nsc.typechecker.Analyzer$namerFactory$$anon$1.apply(Analyzer.scala:48)
        at scala.tools.nsc.Global$GlobalPhase.applyPhase(Global.scala:454)
        at scala.tools.nsc.Global$GlobalPhase.run(Global.scala:401)
        at scala.tools.nsc.Global$Run.compileUnitsInternal(Global.scala:1519)
        at scala.tools.nsc.Global$Run.compileUnits(Global.scala:1503)
        at scala.tools.nsc.Global$Run.compileSources(Global.scala:1495)
        at scala.tools.nsc.Global$Run.compile(Global.scala:1630)
        at scala.tools.nsc.Driver.doCompile(Driver.scala:48)
        at scala.tools.nsc.MainClass.doCompile(Main.scala:30)
        at scala.tools.nsc.Driver.process(Driver.scala:68)
        at scala.tools.nsc.Driver.main(Driver.scala:82)
        at scala.tools.nsc.Main.main(Main.scala)
```

**Comment by SethTisue** [comment]
Thanks, I'm able to reproduce this locally now. (with `libraryDependencies += "org.scala-lang.modules" %% "scala-parser-combinators" % "2.0.0"`)

It happens for me even without `-Ymacro-annotations`, but `-Xsource:3` is necessary.

**Comment by SethTisue** [comment]
Fixed as of 2.13.7-bin-7247dfe, presumably by https://github.com/scala/scala/pull/9669
